### PR TITLE
Re-work determine best networking

### PR DESCRIPTION
**Pull request by lxpollitt**

### CODE REVIEW

**Comment by caseydavenport** [comment]
Deploy preview for *calico* ready!

Built with commit 74b81a9fbce4ea0cedfa41f2de2f5121ae5f5060

https://deploy-preview-3670--calico.netlify.app

**Comment by caseydavenport** [comment]
Deploy preview for *calico* ready!

Built with commit cd797c4eca70fedc89394250211a99ff8c6945d6

https://deploy-preview-3670--calico.netlify.app

File: networking/determine-best-networking.md [hunk 1]
```diff
@@ -1,110 +1,219 @@
...
 
 ### Concepts
 
-#### About {{site.prodname}} networking
+If you want to fully understand the network choices available to you, we recommend you make sure you are familiar with and understand the following concepts.  If you would prefer to skip the learning and get straight to the choices and recommendations, you can jump ahead to [Networking Options](#networking-options).
 
-{{site.prodname}} provides a few ways to allow pods to connect to other pods, to the host, and to outside networks (for example, the internet).
+#### Kubernetes networking basics
+The Kubernetes network model defines a “flat” network in which:
+- every pod get its own IP address
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:20]: Init caps for bullets.

File: networking/determine-best-networking.md [hunk 2]
```diff
@@ -1,110 +1,219 @@
...
+- pods on any node can communicate with all pods on all other nodes without NAT.
 
-{{site.prodname}} networking:
+This creates a clean, backwards-compatible model where Pods can be treated much like VMs or physical hosts from the perspectives of port allocation, naming, service discovery, load balancing, application configuration, and migration.  Network segmentation can be defined using Network Policies to restrict traffic within this base networking capabilities.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:23]: Lowercase for pods and network policies for Concepts...use full resource name in Features, How to section, and how to examples. 

Network segmentation can be defined using network policies to restrict traffic within this base networking capabilities. (these base networking capabilities?)

File: networking/determine-best-networking.md [hunk 3]
```diff
@@ -1,110 +1,219 @@
...
 
-- Assigns IP addresses to pods using {{site.prodname}}’s IP address management (IPAM)
-- Programs the local node’s routing table
-- Distributes routes to other nodes and network devices
+Within this model there’s quite a lot of flexibility for supporting different networking approaches and environments.  The details of exactly how the network is implemented depend on the combination of CNI, network, and cloud-provider plugins being used.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:25]: cloud provider?

File: networking/determine-best-networking.md [hunk 4]
```diff
@@ -1,110 +1,219 @@
...
 
-For details of {{site.prodname}} network design and architecture, see a lightboard presentation on the {{site.prodname}} dataplane, components, and how traffic is routed, {% include open-new-window.html text='Calico network architecture and fundamentals' url='https://www.tigera.io/video/tigera-calico-fundamentals' %}.
+#### CNI plugins
+CNI (Container Network Interface) is a standard API which allows different network implementations to plug into Kubernetes. Kubernetes calls the API any time a pod is being created or destroyed.  There are 2 types of CNI plugins:
+- CNI network plugins: responsible for adding or deleting pods to/from the Kubernetes pod network. This includes creating/deleting each pod’s network interface and connecting/disconnecting it to the rest of the network implementation.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:29]: There are two types...

File: networking/determine-best-networking.md [hunk 5]
```diff
@@ -1,110 +1,219 @@
...
+- CNI IPAM plugins: responsible for allocating and releasing IP addresses for pods as they are created or deleted. Depending on the plugin, this may include allocating one or more ranges of IP addresses (CIDRs) to each node, or obtaining IP addresses from an underlying public cloud’s network to allocate to pods.
 
-#### About BGP
+#### Cloud provider integrations
+Kubernetes cloud provider integrations are cloud specific controllers that can configure the underlying cloud network to help provide Kuberenetes networking.  Depending on the cloud provider, this could include automatically programming routes into the underlying cloud network so it knows natively how to route pod traffic.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:33]: cloud-specific controllers

File: networking/determine-best-networking.md [hunk 6]
```diff
@@ -1,110 +1,219 @@
...
+If you are using an overlay network for your cluster, then pod IPs are not normally routable outside of the cluster.
+
+If you aren’t using an overlay network, then whether pod IPs are routable outside of the cluster depends on what combination of CNI plugins, cloud provider integrations, or (for on-prem) BGP peering with the physical network, is being used.
+
+#### BGP
+BGP (Border Gateway Protocol) is a standards based networking protocol for sharing routes across a network.  It’s one of the fundamental building blocks of the internet, with exceptional scaling characteristics.
+
+{{site.prodname}} has built in support for BGP.  In an on-prem deployment, this allows {{site.prodname}} to peer with the physical network (typically to Top or Rack routers) to exchange routes, making a non-overlay network where pod IP addresses routable across the broader network, just like any other workload attached to the network.
+
+### About {{site.prodname}} Networking
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:82]: About {{site.prodname}} networking (lowercase)

File: networking/determine-best-networking.md [hunk 7]
```diff
@@ -1,110 +1,219 @@
...
+
+{{site.prodname}} can provide both VXLAN or IP-in-IP overlay networks, including cross-subnet only modes.
+
+**Non-overlay network modes**
+
+{{site.prodname}} can provide non-overlay networks running on top of any underlying L2 network, or an L3 network that is either a public cloud network with appropriate cloud provider integration, or a BGP capable network (typically an on-prem network with standard Top-of-Rack routers).
+
+**Network policy enforcement**
+
+{{site.prodname}}’s network policy enforcement engine implements the full range of Kubernetes Network Policy features, plus the extended features of {{site.prodname}} Network Policy.  This works in conjunction with {{site.prodname}}’s built in networking modes, or any other {{site.prodname}} compatible network plugins and cloud provider integrations.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:103]: range of Kubernetes network policy features, plus the extended features of {{site.prodname}} network policy. (lowercase)

File: networking/determine-best-networking.md [hunk 8]
```diff
@@ -1,110 +1,219 @@
...
+
+### {{site.prodname}} compatible CNI plugins and cloud provider integrations
+In addition to the {{site.prodname}} CNI plugins and built in networking modes, {{site.prodname}} is also compatible with a number of third party CNI plugins and cloud provider integrations.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:106]: third-party

File: networking/determine-best-networking.md [hunk 9]
```diff
@@ -1,110 +1,219 @@
...
+
+**Amazon VPC CNI**
+
+The Amazon VPC CNI plugin allocates pod IPs from the underlying AWS VPC and uses AWS elastic network interfaces to provide VPC native pod networking (pod IPs that are routable outside of the cluster). It is the default networking used in {% include open-new-window.html text='Amazon EKS' url='https://aws.amazon.com/eks/' %}, with Calico for network policy enforcement.
 
 **Azure CNI**
 
-The Azure CNI plugin configures the Azure virtual network to provide pod networking. It is the default networking used in Microsoft AKS, with {{site.prodname}} for [network policy enforcement]({{ site.baseurl }}/security/calico-network-policy).
+The Azure CNI plugin allocates pod IPs from the underlying Azure VNET configures the Azure virtual network to provide VNET native pod networking (pod IPs that are routable outside of the cluster). It is the default networking used in {% include open-new-window.html text='Microsoft AKS' url='https://azure.microsoft.com/en-us/services/kubernetes-service/' %}, with Calico for network policy enforcement.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:114]: The Azure CNI plugin allocates pod IPs from the underlying Azure VNET configures the Azure virtual network to provide VNET native pod networking (pod IPs that are routable outside of the cluster). ?? word missing??

File: networking/determine-best-networking.md [hunk 10]
```diff
@@ -1,110 +1,219 @@
...
+
+**Google cloud provider**
+
+The Google cloud provider integration uses host-local IPAM CNI plugin to allocate pod IPs and programs the Google cloud network Alias IP ranges to provide VPC native pod networking on Google cloud (pod IPs that are routable outside of the cluster). It is the default for Google Kubernetes Engine (GKE), with Calico for network policy enforcement.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:118]: The Google cloud provider integration uses host-local IPAM CNI plugin to allocate pod IPs, (comma or it reads "allocate IPs and programs")

{{site.prodname}}

File: networking/determine-best-networking.md [hunk 11]
```diff
@@ -1,110 +1,219 @@
...
+
+**Host local IPAM**
+
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:121]: though two of these...

and {{site.prodname}}

File: networking/determine-best-networking.md [hunk 12]
```diff
@@ -1,110 +1,219 @@
...
+The host local CNI IPAM plugin is a commonly used IP address management CNI plugin, which allocates a fixed size IP address range (CIDR) to each node, and then allocates pod IP addresses from within that range.  The default address range size is 256 IP addresses (a /24), though 2 of those IP addresses are reserved for special purposes and not assigned to pods. The simplicity of host local CNI IPAM plugin makes it easy to understand, but results in less efficient IP address space usage compared to Calico CNI IPAM plugin.
 
 **Flannel**
 
-Flannel routes pod traffic using static per-node CIDRs. It provides a number of networking backends. [Calico can be used for network policy enforcement]({{ site.baseurl }}/security/calico-network-policy).
+Flannel routes pod traffic using static per-node CIDRs obtained from the host-local IPAM CNI plugin. Flannel provides a number of networking backends, but is predominantly used with its VXLAN overlay backend. {{site.prodname}} CNI and {{site.prodname}} network policy can be combined with Flannel and the host-local IPAM plugin to provide a VXLAN network with policy enforcement.  This combination is sometimes referred to as “Canal”.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:126]: Casey convinced me to start using "flannel" lowercase so I've been going with that.

File: networking/determine-best-networking.md [hunk 13]
```diff
@@ -1,110 +1,219 @@
...
+{% include geek-details.html details='Policy:Calico,IPAM:Azure,CNI:Azure,Overlay:No,Routing:VPC Native' %}
 
-Traffic cannot be routed between pods and destinations that aren’t also in the {{site.prodname}} cluster without allowing workload access outside cluster, or peering with infrastructure.
+If you want to use AKS but allocating pod IPs from the underlying VNET is problematic due to IP address range exhaustion challenges, you can use {{site.prodname}} in conjunction with the Azure cloud provider integration. This uses host-local IPAM to allocate /24 per node, and programs routes within the cluster’s underlying VNET subnet for those /24.  Pod IPs are not routable outside of the cluster / VNET subnet, so the same pod IP address range (CIDR) can be used across multiple clusters if desired.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md:168]: "This uses..."  the integration?

File: networking/determine-best-networking.md [hunk 14]
```diff
@@ -1,110 +1,219 @@
...
+
+{% include geek-details.html details='Policy:Calico,IPAM:Calico,CNI:Calico,Cross-subnet:IPIP,Routing:BGP' %}
 
-If possible, we recommend running {{site.prodname}} without network overlay/encapsulation. This provides the highest performance and simplest network; the packet that leaves your workload is the packet that goes on the wire.
+#### Anywhere
 
-However, selectively using overlays (IP in IP or VXLAN) can be useful when running on top of an underlying network that cannot easily be made aware of workload IPs. {{site.prodname}} can perform encapsulation on: all traffic, no traffic, or only on traffic that crosses a subnet boundary.
+The above list of environments is obviously not an exhaustive list. Understanding the concepts and explanations in this guide hopefully will have helped you figure out what is right for your environment. If you still aren't sure then you can ask for advice through the Calico Users's Slack or Discourse forum. And remember you can run Calico in VXLAN overlay mode in almost any environment if you want to get started without worrying too deeply about the different options.
```
Comment (bmckercher123) [inline networking/determine-best-networking.md]: Understanding the concepts and explanations in this guide has hopefully helped you figure out what is right for your environment.

**Comment by caseydavenport** [comment]
^ Approval for technical content, @bmckercher123 should review from a documentation perspective.

File: networking/determine-best-networking.md [hunk 15]
```diff
@@ -1,110 +1,219 @@
...
-{{site.prodname}} can perform network policy enforcement with a number of other Kubernetes networking options.
+The main disadvantages of using an overlay network are:
+- A slight performance impact. The process of encapsulating packets takes a small amount of CPU, and the extra bytes required in the packet to encode the encapsulation (VXLAN or IP-in-IP headers) reduces the maximum size of inner packet that can be sent, which in turn can mean needing to sending more packets for the same amount of total data.
+- The pod IP addresses are not routable outside of the cluster.  More on this below! 
 
-**Amazon AWS VPC CNI**
+#### Cross-subnet overlays
+In addition to standard VXLAN or IP-in-IP overlays, {{site.prodname}} also supports “cross-subnet” modes for VXLAN and IP-in-IP.  In this mode, within each subnet, the underlying network acts as an L2 network. Packets being sent within a single subnet are not encapsulated, so you get the performance of a non-overlay network.  Packets being sent across subnets are encapsulated, like a normal overlay network, reducing dependencies on the underlying network (without the need to integrate with or make any changes to the underlying network).
```
[the remark this answers is not in the document]
Comment (lxpollitt) [reply networking/determine-best-networking.md:48]: Removed the "being".  But I think it should be "reducing" not "reduce".

File: networking/determine-best-networking.md (hunk 10 above)
Comment (lxpollitt) [reply networking/determine-best-networking.md:118]: I don't think this shouldn't be site.prodname.  GKE is integrated with Calico, not Calico Enterprise.  So this stays Calico even in the enterprise docs.

**Pull request merged**